File: implementations/python/packages/mcpproxy/src/psp_cdl_mcpproxy/peer.py
# SPDX-License-Identifier: Apache-2.0
"""Shared pinned discovery and buffered result contract across transports."""
import re
from psp_cdl_core import canonical_json, parse_json
from psp_cdl_api_server.persistence import bounded
from psp_cdl_mcp_server import MCP_VERSION
from psp_cdl_mcp_server.revision import REVISION_PROFILE, REVISION_KEY, valid_catalog_revision
from psp_cdl_api_server.service import identifier
from .dispatch import binding_digest

class PeerError(ValueError):
    def __init__(self, code):
        super().__init__(code)
        self.code = code

def json_copy(value):
    try: return bounded(value)
    except Exception: raise PeerError("INVALID_PEER_DATA") from None

class PinnedMcpClient:
    @property
    def catalog_digest(self): return self._fingerprint

    @property
    def catalog_snapshot(self): return json_copy({"tools":self._catalog,"revision":self._revision,"approvalDigest":self._fingerprint})

    def _digest(self,view): return binding_digest(view if self._required_revision else view["tools"])

    def _initialize(self, info, profile=None):
        self._required_revision=profile==REVISION_PROFILE
        init = self._request("initialize", {"protocolVersion":MCP_VERSION, "capabilities":{"experimental":{REVISION_KEY:{"profile":REVISION_PROFILE}}} if self._required_revision else {}, "clientInfo":{"name":"psp-cdl-mcpproxy", "version":"0.1.0"}})
        if type(init) is not dict or init.get("protocolVersion") != MCP_VERSION or type(init.get("serverInfo")) is not dict or any(init["serverInfo"].get(k) != info[k] for k in ("name", "version")) or type(init.get("capabilities")) is not dict or type(init["capabilities"].get("tools")) is not dict:
            raise PeerError("INVALID_INITIALIZATION")
        experimental=init["capabilities"].get("experimental")
        if self._required_revision and (type(experimental) is not dict or experimental.get(REVISION_KEY)!={"profile":REVISION_PROFILE}): raise PeerError("REVISION_UNSUPPORTED")
        self._notify("notifications/initialized", {})
        view=self._discover()
        self._catalog,self._revision=view["tools"],view["revision"]
        self._fingerprint=self._digest(view)

    def _discover(self, cancelled=lambda:False):
        result = self._request("tools/list", {}, cancelled)
        if type(result) is not dict or set(result) != ({"tools","_meta"} if self._required_revision else {"tools"}) or type(result["tools"]) is not list or len(result["tools"]) > 1024: raise PeerError("INVALID_DISCOVERY")
        seen = set()
        for t in result["tools"]:
            if type(t) is not dict or type(t.get("name")) is not str or not re.fullmatch(r"[A-Za-z0-9_-]{1,64}", t["name"]) or t["name"] in seen or type(t.get("inputSchema")) is not dict or type(t.get("outputSchema")) is not dict: raise PeerError("INVALID_DISCOVERY")
            seen.add(t["name"])
        tools=json_copy(sorted(result["tools"],key=lambda t:t["name"]))
        revision=None
        if self._required_revision:
            meta=result.get("_meta")
            r=meta.get(REVISION_KEY) if type(meta) is dict else None
            if type(meta) is not dict or set(meta)!={REVISION_KEY} or not valid_catalog_revision(r) or r["catalogDigest"]!=binding_digest(tools): raise PeerError("INVALID_REVISION_DATA")
            for t in tools:
                m=t.get("_meta")
                if type(m) is not dict or type(m.get(REVISION_KEY)) is not dict or set(m[REVISION_KEY])!={"toolRevision"} or not identifier(m[REVISION_KEY]["toolRevision"]): raise PeerError("INVALID_REVISION_DATA")
            revision=json_copy(r)
        return {"tools":tools,"revision":revision}

    def registrations(self, server, approvals, now, approved_catalog_digest=None):
        if self._required_revision and approved_catalog_digest!=self._fingerprint: raise PeerError("CATALOG_NOT_APPROVED")
        approvals = json_copy(approvals)
        if type(approvals) is not list or not callable(now): raise PeerError("INVALID_CONFIGURATION")
        result, seen = [], set()
        for a in approvals:
            t = next((t for t in self._catalog if t["name"] == a.get("name")), None)
            if t is None or a["name"] in seen or any(canonical_json(t[k]) != canonical_json(a.get(k)) for k in ("inputSchema", "outputSchema")): raise PeerError("DISCOVERY_MISMATCH")
            if self._required_revision and a["revision"]!=t["_meta"][REVISION_KEY]["toolRevision"]: raise PeerError("DISCOVERY_MISMATCH")
            seen.add(a["name"])
            def invoke(args, options, name=a["name"], tool_revision=a["revision"]):
                cancelled = lambda:options["cancelled"]() or now() >= options["deadline"]
                try:
                    if self._digest(self._discover(cancelled)) != self._fingerprint: raise PeerError("DISCOVERY_CHANGED")
                    pre={**self._revision,"toolRevision":tool_revision,"inputDigest":binding_digest(args)} if self._required_revision else None
                    response = self._request("tools/call", {"name":name, "arguments":args,**({"_meta":{REVISION_KEY:pre}} if pre else {})}, cancelled)
                    if type(response) is not dict or set(response)-{"content","structuredContent","isError","_meta"} or ("isError" in response and response["isError"] is not False) or type(response.get("structuredContent")) is not dict or type(response.get("content")) is not list or len(response["content"]) > 1: raise PeerError("INVALID_TOOL_RESPONSE")
                    if pre and (type(response.get("_meta")) is not dict or canonical_json(response["_meta"].get(REVISION_KEY))!=canonical_json(pre)): raise PeerError("INVALID_REVISION_RECEIPT")
                    for block in response["content"]:
                        if type(block) is not dict or set(block) != {"type","text"} or block["type"] != "text" or type(block["text"]) is not str or canonical_json(parse_json(block["text"])) != canonical_json(response["structuredContent"]): raise PeerError("INVALID_TOOL_RESPONSE")
                    if self._digest(self._discover(cancelled)) != self._fingerprint: raise PeerError("DISCOVERY_CHANGED")
                    return json_copy(response["structuredContent"])
                except Exception:
                    self.close()
                    raise
            result.append({**a, "server":server, "invoke":invoke})
        return result

File: implementations/typescript/packages/mcpproxy/src/peer.ts
// SPDX-License-Identifier: Apache-2.0
import {canonicalJson,parseJson,record} from "@psp-cdl/core";
import {MCP_VERSION} from "@psp-cdl/mcp-server";
import {bounded} from "@psp-cdl/api-server/persistence";
import {type ToolRegistration,bindingDigest} from "./dispatch.js";
export class PeerError extends Error {constructor(public readonly code:string){super(code);this.name="PeerError";}}
type Approval=Omit<ToolRegistration,"server"|"invoke">;
const fail=(code:string):never=>{throw new PeerError(code);};
export function json(value:unknown):any {try{return bounded(value);}catch{return fail("INVALID_PEER_DATA");}}
/** Shared pinned discovery and buffered tool-result contract across transports. */
export abstract class PinnedMcpClient {
  private catalog:Record<string,unknown>[]=[];
  private fingerprint="";
  get catalogDigest():string{return this.fingerprint;}
  protected abstract request(method:string,params:unknown,cancelled?:()=>boolean):Promise<any>;
  protected abstract notify(method:string,params:unknown):Promise<void>;
  abstract close():Promise<void>;
  protected async initialize(info:{name:string;version:string}):Promise<void> {
      const init=await this.request("initialize",{protocolVersion:MCP_VERSION,capabilities:{},clientInfo:{name:"psp-cdl-mcpproxy",version:"0.1.0"}});
      if(!record(init)||init.protocolVersion!==MCP_VERSION||!record(init.serverInfo)||init.serverInfo.name!==info.name||init.serverInfo.version!==info.version||!record(init.capabilities)||!record(init.capabilities.tools)) fail("INVALID_INITIALIZATION");
      await this.notify("notifications/initialized",{});
      this.catalog=await this.discover();this.fingerprint=bindingDigest(this.catalog);
  }
  private async discover(cancelled?:()=>boolean):Promise<Record<string,unknown>[]> {
    const result=await this.request("tools/list",{},cancelled);
    if(!record(result)||Object.keys(result).some(k=>k!=="tools")||!Array.isArray(result.tools)||result.tools.length>1024)fail("INVALID_DISCOVERY");
    const seen=new Set<string>();
    for(const t of result.tools) {
      if(!record(t)||typeof t.name!=="string"||!/^[A-Za-z0-9_-]{1,64}$/.test(t.name)||seen.has(t.name)||!record(t.inputSchema)||!record(t.outputSchema))fail("INVALID_DISCOVERY");
      seen.add(t.name);
    }
    return json([...result.tools].sort((a,b)=>a.name<b.name?-1:1));
  }
  /** Only explicit host approvals establish capabilities/read-only admission. */
  registrations(server:string,approvals:Approval[],now:()=>number):ToolRegistration[] {
    const values=json(approvals) as Approval[],seen=new Set<string>();
    if(!Array.isArray(values)||typeof now!=="function")fail("INVALID_CONFIGURATION");
    return values.map(a=>{
      const t=this.catalog.find(t=>t.name===a.name);
      if(!t||seen.has(a.name)||canonicalJson(t.inputSchema)!==canonicalJson(a.inputSchema)||canonicalJson(t.outputSchema)!==canonicalJson(a.outputSchema))fail("DISCOVERY_MISMATCH");
      seen.add(a.name);
      return {...a,server,invoke:async(args,options)=>{
        const cancelled=()=>options.cancelled()||now()>=options.deadline;
        try {
          if(bindingDigest(await this.discover(cancelled))!==this.fingerprint)fail("DISCOVERY_CHANGED");
          if(cancelled())fail("PEER_CANCELLED");
          const response=await this.request("tools/call",{name:a.name,arguments:args},cancelled);
          if(!record(response)||Object.keys(response).some(k=>!["content","structuredContent","isError","_meta"].includes(k))||response.isError!==false&&response.isError!==undefined||!record(response.structuredContent)||!Array.isArray(response.content)||response.content.length>1)fail("INVALID_TOOL_RESPONSE");
          for(const block of response.content) {
            if(!record(block)||Object.keys(block).sort().join(",")!=="text,type"||block.type!=="text"||typeof block.text!=="string"||canonicalJson(parseJson(block.text))!==canonicalJson(response.structuredContent))fail("INVALID_TOOL_RESPONSE");
          }
          if(bindingDigest(await this.discover(cancelled))!==this.fingerprint)fail("DISCOVERY_CHANGED");
          return json(response.structuredContent);
        }catch(e){await this.close();throw e;}
      }};
    });
  }
}
